src/transform/index.ts

```diff
@@ -0,0 +1,35 @@
+import { spawn } from 'duplex-child-process';
+import { Readable } from 'stream';
+import { Operation } from '../types/Operation';
+import ImageDefinition from './../imagedef';
+import analyze from './../imagedef/analyze';
+import StreamSwitch from './../lib/stream-switch';
+import initializePipeline from './initialize-pipeline';
+import joinCommands from './join-commands';
+import simulateTransformation from './simulate-transformation';
+
+
+export default async (stream: Readable, steps: Operation[]): Promise<Readable> => {
+  if (steps.length === 0) {
+    return stream;
+  }
+
+  const streamSwitch = new StreamSwitch(stream);
+  const streamToAnalyze = streamSwitch.createReadStream();
+  const streamToTransform = streamSwitch.createReadStream();
+
+  // initialize steps
+  const { pipeline, requirements } = initializePipeline(steps);
+
+  // prepare the image definition
+  const definition: ImageDefinition = await analyze(streamToAnalyze, requirements);
+
+  // build command from pipeline and image state
+  const { commands } = simulateTransformation(pipeline, definition);
```

Comment:
why do we simulate the transformation here?

Reply:
Perhaps the name is misleading, what this does is simply to go through the operations defined in the pipeline and transform the image state. The result is the commands and final state.